### PR TITLE
chore(WIP): add architecture readme

### PR DESCRIPTION
**Pull request by enricopolanski**
I wanted to recheck what I wrote few weeks ago when analyzing the architecture.

I think this is a nice starting point, but I feel like half through the list of components I keep giving definitions and examples but don't really tie it to the `effect-ts-app` boilerplate anymore.

What I'd like to do before merging it, is to tie it more closely to the boilerplate architecture trying to show the various components definitions against the boilerplate itself and it's structure.

Any feedback is more than welcome.

### CODE REVIEW

File: Architecture.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+Another application used in a warehouse may handle the inventory, packaging and operations with the carrier. But it may neither have the capabilities to create or review orders, nor to handle the payments or communication with the customer.
+
+## Controller
+
+A `Controller` is a piece of application-specific logic. A `Conteoller` defines a single capability of the application such as creating `Order`s. `Controller`s use the logic defined in `Core` and use injected dependencies (through inversion of control) to handle communication with external `Service`s.
+
+## Service
+
+A `Service` provides a dependency needed by a `Controller`.
```
Comment (patroza) [inline Architecture.md]: Services are more general. Controllers consume services for sure, but Services can be used anywhere.
Generally the type of Service dictates where it should be used.
An infrastructure service is used in Controllers, but not in domain methods etc.

File: Architecture.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+## Repository
+
+A `Repository` is a `Service` that provides APIs for CRUD operations for some `Entity` while hiding the actual implementation and driver-level details.
+
+An example of a `Repository` is a for our `Order` `Entity` would be an `OrderRepository` which would provide the necessary APIs to perform CRUD operations involving `Order` `Entity`s.
+
+## Resource
+
+A `Resource` provides specific `Request/Response` messages that `Controller`s take as input or return as output.
```
Comment (patroza) [inline Architecture.md]: And `Client`s use them to derive fully typed api cleints.

**Comment by patroza** [comment]
Let's rename `effect-ts-app` to `effect-app`

File: Architecture.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+By defining `Entity`s encapsulations we can define what `Order`s are (their `Schema`) and how, when and what can be done with them (through their `Core` logic). The business rules encoded in `Core` such as the inability to place `Order`s with a delivery date set in the past, are general and independent from the class of applications that may interface with the `Order`s.
+
+## Value-Object
+
+Similar to `Entity`s, `Value-Object`s also abstract a _business domain object_. The core difference between `Entity`s and `Value-Object`s is that `Value-Object`s do not have a unique identifier.
+`Value-Object`s too are defined by their data shape and their business logic, and those too are defined through `Schema` and `Core` logic inthe same packages where `Entity`s are defined.
+
+An example of a `Value-Object` might be an email or a geographical `Address`. An `Order` might have a delivery `Address` but this `Address` is only defined by its content. `Value-Object`s are generally attached to `Entity`s and modifications of `Value-Object`s are indistinguishable from replacing the `Value-Object`s as they do not hold any unique identifier and their identity is defined by their content only.
```
Comment (patroza) [inline Architecture.md:61]: an Email is generally a refined primitive string, not a value **Object**.
We should add a section about such Refinements.

Comment (enricopolanski) [reply Architecture.md:61]: So what would be an example of a `Value-Object`?

Comment (patroza) [reply Architecture.md:61]: @enricopolanski usually Address, defined as e.g addressfield1, 2, 3, street, number, zip, etc.

**Pull request merged**